shared/src/balancer/swap/fixed_point.rs

```diff
@@ -1,3 +1,307 @@
+//! Module emulating the operations on fixed points with exactly 18 decimals as
+//! used in the Balancer smart contracts. Their original implementation can be
+//! found at:
+//! https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/solidity-utils/contracts/math/FixedPoint.sol
+
 #![allow(dead_code)]
 
+use super::error::Error;
+use anyhow::{anyhow, bail};
+use ethcontract::U256;
+use lazy_static::lazy_static;
+use std::{
+    fmt::{self, Debug, Formatter},
+    str::FromStr,
+};
+
 mod logexpmath;
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+/// Fixed point numbers that represent exactly any rational number that can be
+/// represented with up to 18 decimals as long as it can be stored in 256 bits.
+/// It corresponds to Solidity's `ufixed256x18`.
+/// Operations on this type are implemented as in Balancer's FixedPoint library,
+/// including error codes, from which the name (Balancer Fixed Point).
+pub struct Bfp(U256);
+
+lazy_static! {
+    static ref ONE_18: U256 = U256::exp10(18);
+    static ref ZERO: Bfp = Bfp(U256::zero());
+    static ref EPSILON: Bfp = Bfp(U256::one());
+    static ref ONE: Bfp = Bfp(*ONE_18);
+    static ref MAX_POW_RELATIVE_ERROR: Bfp = Bfp(10000_usize.into());
+}
+
+impl From<usize> for Bfp {
+    fn from(num: usize) -> Self {
+        Self(U256::from(num).checked_mul(*ONE_18).unwrap())
+    }
+}
+
+impl FromStr for Bfp {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut split_dot = s.split('.');
+        let units = split_dot.next().unwrap();
+        let decimals = split_dot.next().unwrap_or("0");
+        if split_dot.next().is_some() {
+            bail!("Too many decimal separators");
+        }
+        if units.is_empty() || decimals.is_empty() || decimals.len() > 18 {
+            bail!("Invalid decimal representation");
+        }
+        Ok(Bfp(U256::from_dec_str(&format!("{:0<18}", decimals))?
+            .checked_add(
+                U256::from_dec_str(units)?
+                    .checked_mul(*ONE_18)
+                    .ok_or_else(|| anyhow!("Too large number"))?,
+            )
+            .ok_or_else(|| anyhow!("Too large number"))?))
+    }
+}
+
+impl Debug for Bfp {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(
+            formatter,
+            "{}.{:0>18}",
+            self.0 / *ONE_18,
+            (self.0 % *ONE_18).as_u128()
+        )
+    }
+}
+
+impl Bfp {
+    pub const MAX: Self = Self(U256::MAX);
+
+    pub fn as_uint256(self) -> U256 {
+        self.0
+    }
+
+    pub fn zero() -> Self {
+        *ZERO
+    }
+
+    pub fn one() -> Self {
+        *ONE
+    }
+
+    pub fn epsilon() -> Self {
+        *EPSILON
+    }
+
+    pub fn from_wei(num: U256) -> Self {
+        Self(num)
+    }
```

Comment:
I would expect the `from_wei` is at least a bit more involved than this.

Reply:
This function exists because I don't want to make the constructor public. What I think could be an issue with that is the confusion that creating something like `Bfp(100)` is not going to create 100.0, but 10^-16. So to make it super clear that decimals are involved I called the only external constructor `from_wei`, just because luckily it happens that Balancer picked 18 decimals for their fixed point representation. So it's really that easy! :smile: